packages/zoe/src/contracts/newSwap/swap.js

```diff
@@ -0,0 +1,262 @@
+          pool.getCentralAmount(),
+          amountMath.add(amountOut, protocolFee),
+        ),
+        Secondary: amountMath.add(pool.getSecondaryAmount(), reducedAmountIn),
+      });
+    }
+
+    function secondaryPoolStagings() {
+      const brandInPool = getPool(brandIn);
+      // gains reducedAmountIn, loses (centralAmount + protocolFee)
+      const brandInStaging = brandInPool.stageSeat({
+        Secondary: amountMath.add(
+          brandInPool.getSecondaryAmount(),
+          reducedAmountIn,
+        ),
+        Central: amountMath.subtract(
+          brandInPool.getCentralAmount(),
+          amountMath.add(centralAmount, protocolFee),
+        ),
+      });
+
+      const brandOutPool = getPool(brandOut);
+      // gains centralAmount, loses amountOut
+      const brandOutStaging = brandOutPool.stageSeat({
+        Central: amountMath.add(brandOutPool.getCentralAmount(), centralAmount),
+        Secondary: amountMath.subtract(
+          brandOutPool.getSecondaryAmount(),
+          amountOut,
+        ),
+      });
+      return [brandInStaging, brandOutStaging];
+    }
+
+    // central to secondary, secondary to central, or secondary to secondary
+    const pools = [];
+    if (isCentral(brandIn) && isSecondary(brandOut)) {
+      stagings.push(poolStagingFromCentral());
+      pools.push(getPool(brandOut));
+    } else if (isSecondary(brandIn) && isCentral(brandOut)) {
+      stagings.push(poolStagingToCentral());
+      pools.push(getPool(brandIn));
+    } else if (isSecondary(brandIn) && isSecondary(brandOut)) {
+      stagings.push(...secondaryPoolStagings());
+      pools.push(getPool(brandIn));
+      pools.push(getPool(brandOut));
+    } else {
+      assert.fail(X`brands were not recognized`);
+    }
+
+    // @ts-ignore typescript can't tell that we started by adding two stagings
+    zcf.reallocate(...stagings);
+    seat.exit();
+    pools.forEach(p => p.updateState());
+    return SUCCESS_STRING;
+  };
+
+  // trade with a stated amount out.
+  const swapOut = seat => {
+    assertProposalShape(seat, {
+      give: { In: null },
+      want: { Out: null },
+    });
+    // The offer's amountOut is a minimum; the offeredAmountIn is a max.
+    const {
+      give: { In: offeredAmountIn },
+      want: { Out: amountOut },
+    } = seat.getProposal();
+    const { brand: brandOut } = amountOut;
+    const brandIn = offeredAmountIn.brand;
+
+    const {
+      amountIn,
+      amountOut: improvedAmountOut,
+      protocolFee,
+      // @ts-ignore two pool version of getPrice returns central
+      centralAmount,
+    } = getPriceGivenRequiredOutput(brandIn, amountOut);
+
+    const stagings = [];
+    stagings.push(stageProtocolSeatFee(protocolFee));
+    // The seat's staging is the same for every trade
+    stagings.push(
+      seat.stage({
+        In: amountMath.subtract(offeredAmountIn, amountIn),
+        Out: improvedAmountOut,
+      }),
+    );
+
+    function poolStagingToCentral() {
```

Comment:
This is almost identical to the code above.  It could be a shared function of 2 arguments, the `amountOut` and the `amoutnIn`.  Not sure whether this would be a clarifying improvement, but it might reveal further cleanups.

Reply:
It needed three parameters, (protocolFee, too), but the same was true of the other two parallel functions also.